src/v2/cookbook/avoiding-memory-leaks.md

```diff
@@ -0,0 +1,169 @@
+---
+title: Avoiding Memory Leaks
+type: cookbook
+order: 9
+---
+## Introduction
+
+If you are developing applications with Vue, then you need to watch out for memory leaks. This issue is especially important in Single Page Applications (SPAs) because by design, users should not have to refresh their browser when using an SPA, so it is up to the Javascript application to clean up components and make sure that garbage collection takes place as expected.
+
+Memory leaks in Vue applications do not typically come from Vue itself, rather they can happen when incorporating other libraries into an application.
+
+## Simple Example
+
+The following example shows a memory leak caused by using the [Choices.js](https://github.com/jshjohnson/Choices) library in a Vue component and not properly cleaning it up. Later, we will show how to remove the Choices.js footprint and avoid the memory leak.
+
+In the example below, we load up a select with a lot of options and then we use a show/hide button with a [v-if](/v2/guide/conditional.html) directive to add it and remove it from the virtual DOM. The problem with this example is that the v-if directive removes the parent element from the DOM, but we did not clean up the additional DOM pieces created by Choices.js, causing a memory leak. 
+
+```html
+<link rel='stylesheet prefetch' href='https://joshuajohnson.co.uk/Choices/assets/styles/css/choices.min.css?version=3.0.3'>
+<script src='https://joshuajohnson.co.uk/Choices/assets/scripts/dist/choices.min.js?version=3.0.3'></script>
+
+<div id="app">
+  <button v-if="showChoices" @click="hide">Hide</button>
+  <button v-if="!showChoices" @click="show">Show</button>
+  <div v-if="showChoices">
+    <select id="choices-single-default"></select>
+  </div>
+</div>
+```
+```js
+new Vue({
+  el: "#app",
+  data: function () {
+    return {
+      showChoices: true
+    }
+  },
+  mounted: function () {
+    this.initializeChoices()
+  },
+  methods: {
+    initializeChoices: function () {
+      let list = []
+      // lets load up our select with many choices 
+      // so it will use a lot of memory
+      for (let i = 0; i < 1000; i++) {
+        list.push({
+          label: "Item " + i,
+          value: i
+        })
+      }
+      new Choices("#choices-single-default", {
+        searchEnabled: true,
+        removeItemButton: true,
+        choices: list
+      })
+    },
+    show: function () {
+      this.showChoices = true
+      this.$nextTick(() => {
+        this.initializeChoices()
+      })
+    },
+    hide: function () {
+      this.showChoices = false
+    }
+  }
+})
+```
+To see this memory leak in action, open this [CodePen example](https://codepen.io/freeman-g/pen/qobpxo) using Chrome and then open the Chrome Task Manager (Chrome Menu > More Tools > Task Manager). Now, click the show/hide button 50 or so times. You should see the memory usage in the Chrome Task Manager increase and never be reclaimed.
```

Comment:
Huh, this is not located here for me. For me it's Chrome top navigation > Window > Task Manager. I'm on version 66.0.3359.117 Mac, US. If it's different for you, we could always include both and just say it's either here or here.

Reply:
Hi - I also have it under Window > Task manager. I can change to this since its easier to find there.

On Windows, there is a Shift-Esc shortcut to open the Task Manager. I can't find a mac shortcut.

Should I explain both the windows and mac approaches? Something like this?

```
...then open the Chrome Task Manager. To open the Chrome Task Manager on Mac, choose Chrome Top Navigation > Window > Task Manager or on Windows, use the Shift+Esc shortcut.
```

Reply:
Yes that's perfect 👍